operator/src/test/java/org/keycloak/operator/KeycloakDeploymentE2EIT.java

```diff
@@ -0,0 +1,138 @@
+        } catch (Exception e) {
+            savePodLogs();
+            throw e;
+        }
+    }
+
+    @Test
+    public void testImageChange() {
+        try {
+            var kc = getDefaultKeycloakDeployment(k8sclient);
+            kc.getSpec().setImage("quay.io/keycloak/non-existing-keycloak");
+
+            deployKeycloak(k8sclient, kc, false);
+            waitForKcToFail(kc);
+        } catch (Exception e) {
+            savePodLogs();
+            throw e;
+        }
+    }
+
+    @Test
+    public void testServerConfiguration() {
+        try {
+            var kc = getDefaultKeycloakDeployment(k8sclient);
+            kc.getSpec().getServerConfiguration().put("KC_DB_PASSWORD", "Ay Caramba!");
+
+            deployKeycloak(k8sclient, kc, false);
+            waitForKcToFail(kc);
+        } catch (Exception e) {
+            savePodLogs();
+            throw e;
+        }
+    }
+
+    @Test
+    public void testDeploymentDurability() throws InterruptedException {
+        try {
+            var kc = getDefaultKeycloakDeployment(k8sclient);
+            var deploymentName = kc.getMetadata().getName();
+            deployKeycloak(k8sclient, kc, true);
+
+            Log.info("Trying to delete deployment");
+            assertThat(k8sclient.apps().deployments().withName(deploymentName).delete()).isTrue();
+            Awaitility.await()
+                    .untilAsserted(() -> assertThat(k8sclient.apps().deployments().withName(deploymentName).get()).isNotNull());
+
+            waitForKeycloakToBeReady(k8sclient, kc); // wait for reconciler to calm down to avoid race condititon
+
+            Log.info("Trying to modify deployment");
+
+            var deployment = k8sclient.apps().deployments().withName(deploymentName).get();
+            var labels = Map.of("address", "EvergreenTerrace742");
+            var flandersEnvVar = new EnvVarBuilder().withName("NEIGHBOR").withValue("Stupid Flanders!").build();
+            var origSpecs = deployment.getSpec();
+
+            deployment.getMetadata().getLabels().putAll(labels);
+            deployment.getSpec().getTemplate().getSpec().getContainers().get(0).setEnv(List.of(flandersEnvVar));
+            k8sclient.apps().deployments().createOrReplace(deployment);
+
+            Awaitility.await()
+                    .untilAsserted(() -> {
+                        var d = k8sclient.apps().deployments().withName(deploymentName).get();
+                        assertThat(d.getMetadata().getLabels().entrySet().containsAll(labels.entrySet())).isTrue(); // additional labels should not be overwritten
+                        assertThat(d.getSpec()).isEqualTo(origSpecs); // specs should be reconciled back to original values
+                    });
+        } catch (Exception e) {
+            savePodLogs();
+            throw e;
+        }
+    }
+
+    // TODO Test HasErrors condition; ReplicaFailure is not easily reproducible
```

Comment:
Given this comment, should we iterate over this logic? (in a separate ticket)
https://github.com/keycloak/keycloak/blob/011d108fff1554e1784e0b9c70552e45ded7ec50/operator/src/main/java/org/keycloak/operator/v2alpha1/KeycloakDeployment.java#L144-L150